pkg/upstream/cluster/strict_dns_cluster.go

```diff
@@ -0,0 +1,311 @@
+			log.DefaultLogger.Debugf("[upstream] [strict dns cluster] create a resolver for address: %s", rt.dnsAddress)
+		}
+	}
+	sdc.resolveTargets = rts
+}
+
+func (sdc *strictDnsCluster) StopResolve() {
+	for _, rt := range sdc.resolveTargets {
+		rt.StopResolve()
+	}
+}
+
+// calculate next resolve interval
+// 1. if respectDnsTTL configured, and minTtl > 0, use minTtl+1
+// 2. if dnsRefreshRate configured, use dnsRefreshRate
+// 3. use DefaultRefreshInterval(5s)
+func (sdc *strictDnsCluster) calculateNextResolveInterval(minTtl time.Duration) time.Duration {
+	dnsRefreshRate := sdc.dnsRefreshRate
+	if sdc.respectDnsTTL && minTtl > 0 {
+		// increase minTtl by 1 in case we will get dns response with ttl 0
+		dnsRefreshRate = minTtl + time.Second
+	} else if sdc.dnsRefreshRate == 0 {
+		dnsRefreshRate = DefaultRefreshInterval
+	}
+	return dnsRefreshRate
+}
+
+// just check address string, hostname and metadata
+func hostEqual(hosts1, hosts2 *[]types.Host) bool {
+	hosts := map[string]types.Host{}
+	if len(*hosts1) != len(*hosts2) {
+		return false
+	}
+	for _, host := range *hosts1 {
+		hosts[host.AddressString()] = host
+	}
+	for _, host := range *hosts2 {
+		if h, exists := hosts[host.AddressString()]; exists {
+			if h.Hostname() == host.Hostname() && reflect.DeepEqual(h.Metadata(), host.Metadata()) {
+				continue
+			}
+			return false
+		}
+		return false
+	}
+	return true
+}
+
+// combine hosts in current rt and other rt
+func (sdc *strictDnsCluster) updateDynamicHosts(newHosts []types.Host, rt *ResolveTarget) {
+	sdc.mutex.Lock()
+	defer sdc.mutex.Unlock()
+
+	// update current hosts in rt
+	rt.hosts = newHosts
+
+	var allHosts []types.Host
+	// collect all hosts in resolve targets
+	for _, r := range sdc.resolveTargets {
+		for _, h := range r.hosts {
+			allHosts = append(allHosts, h)
+		}
+	}
+
+	// compare current hosts with allHosts
+	hostNotChanged := hostEqual(&allHosts, &sdc.hostSet.allHosts)
+	if !hostNotChanged {
+		for _, h := range newHosts {
+			if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
+				log.DefaultLogger.Infof("[upstream] [strict dns cluster] resolve dns new address:%s", h.AddressString())
+			}
+		}
+		sdc.simpleCluster.UpdateHosts(newHosts)
+		if log.DefaultLogger.GetLogLevel() >= log.INFO {
+			log.DefaultLogger.Infof("[upstream] [strict dns cluster] resolve dns result updated, cluster_name:%s, address:%s", sdc.simpleCluster.info.name, rt.dnsAddress)
+		}
+	}
+}
+
+func (rt *ResolveTarget) StopResolve() {
+	close(rt.stop)
+}
+
+func (rt *ResolveTarget) StartResolve() {
+	defer func() {
+		if r := recover(); r != nil {
+			if log.DefaultLogger.GetLogLevel() >= log.ERROR {
+				log.DefaultLogger.Errorf("[upstream] [strict_dns_cluster] [resolver] panic %v\n%s", r, string(debug.Stack()))
+			}
+		}
+		rt.resolveTimer.Stop()
+		rt.resolveTimeout.Stop()
+	}()
+
+	// start resolve now
+	rt.resolveTimer = utils.NewTimer(0, rt.OnResolve)
+	for {
+		select {
+		case <-rt.stop:
```

Comment:
Should be call `rt.resolveTimer.Stop()` , Because `UpdateClusterHosts()` and `OnResolve()` may be called by multiple goroutines, this will cause overwriting problems with data updates.  In addition, we should give priority to the `rt.stop` event, forexample：
```
for {
    select {
    case <-rt.stop:
        xxx
    default:
        select {
        case <-rt.stop:
            xxx
        case rt.timeout:
           xxx
        }
    }
}
```